### PR TITLE
Gui refactor

### PR DESCRIPTION
**Pull request by ryanhaas**
---
name: Pull request
about: Add changes to the current project
title: Gui refactor
lables: bug fix, clean up, actions
asignees: Ryan, Dylan

---

**What is the pull request for?**
Fixed the GUI tests to run in a headless environment, automatic creation of coverage reports.

**Describe what the new code does**
Added several mock classes that can do the bare minimum functions to emulate the Java Swing/AWT objects to allow testing in a headless environment with no GUI. Updated the tests to reflect these changes (had to change some casting and fix misc. issues). Also created a new plugin to the pom.xml called JaCoco which will automatically generate coverage reports. Created a new Git workflow that automatically publishes these reports to codecov.io

### CODE REVIEW

File: src/main/java/views/components/GUIClass.java [hunk 1]
```diff
@@ -49,11 +52,25 @@
...
 	private JPanel methodRegion;
 	
+	private boolean human;
+	
 	/**
 	 * Initialize a graphical view of a given UMLClass
 	 * @param umlClass
 	 */
 	public GUIClass(UMLClass umlClass) {
+		human = false;
```
Comment (willkill07) [inline src/main/java/views/components/GUIClass.java:62]: should maybe call super(false); ?

File: src/main/java/views/components/GUIClass.java [hunk 2]
```diff
@@ -126,8 +145,8 @@ public void setLocation(int x, int y) {
 	 * @param text - The label text
 	 * @return - JLabel instance
 	 */
-	public static JLabel generateLabel(String text) {
-		JLabel temp = new JLabel(text);
+	public JLabel generateLabel(String text) {
+		JLabel temp = (human) ? new JLabel(text) : new TestableLabel(text);
```
Comment (willkill07) [inline src/main/java/views/components/GUIClass.java:149]: This is an awesome case for a builder design pattern... e.g.

```java
JLabel temp = new LabelBuilder().setText("text").setHuman(human).build()
```

File: src/main/java/views/components/GUIClass.java [hunk 3]
```diff
@@ -137,8 +156,8 @@ public static JLabel generateLabel(String text) {
 	 * Generate a horizontal separator
 	 * @return - JSeparator instance
 	 */
-	public static JSeparator generateSeparator(Color c) {
-		JSeparator mySep = new JSeparator(SwingConstants.HORIZONTAL);
+	public JSeparator generateSeparator(Color c) {
+		JSeparator mySep = (human) ? new JSeparator(SwingConstants.HORIZONTAL) : new TestableSeparator();
```
Comment (willkill07) [inline src/main/java/views/components/GUIClass.java:160]: Same here 🙃

**Review by willkill07 — COMMENTED** [review]
Pretty impressed with all of this. Minor feedback (though not required)

**Review by anthony81799 — APPROVED** [review]
Looks good.  Nice use of mock joption with the testable classes.

**Pull request merged**